### PR TITLE
Monkey cap done right

### PR DESCRIPTION
**Pull request by ghost**
:cl: More Robust Than You
config: You can now set a maximum monkey cap via the config
/:cl:

Closes #31678

### CODE REVIEW

**Comment by PKPenguin321** [comment]
Defaults to infinity, right?

**Comment by GunHog** [comment]
Why do we need a cap for monkeys?

**Comment by Supermichael777** [comment]
does this count player controlled monkeys?

**Comment by Supermichael777** [comment]
`150 is a lot
might want to limit it closer to 40 in practice, though if it counts dead monkeys you could grief xeno by spacing a bunch. The real problem is that restricted hydro machinery got put in all access

**Comment by Shdorsh** [comment]
Thanks mate. I saw that the script I took accidentally was from the tgstation branch, so it removed some of the stuff that was added. Also, if it is the same function I had, it usually is only triggered on monkeys spawned by cubes, any other monkey that goes over the cap is fine since else, it might interfere with gameplay like genetics not being able to turn into monkeys or diseases that turn players into monkeys and such.

**Comment by MoreRobustThanYou** [comment]
@Shdorsh I made it so it won't check monkeys with ckeys

File: code/controllers/configuration/entries/game_options.dm [hunk 1]
```diff
@@ -269,3 +269,7 @@ CONFIG_DEF(number/emergency_shuttle_autocall_threshold)
 	min_val = 0
 	max_val = 1
 	integer = FALSE
+
+CONFIG_DEF(number/monkeycap) //how many monkeys should we allow to exist at once
+	value = 150
```
Comment (Cyberboss) [inline code/controllers/configuration/entries/game_options.dm]: Default value should be INFINITY

File: code/modules/mob/living/carbon/monkey/monkey.dm [hunk 2]
```diff
@@ -35,6 +43,10 @@
 	create_dna(src)
 	dna.initialize_dna(random_blood_type())
 
+/mob/living/carbon/monkey/Destroy()
+	. = ..()
```
Comment (Cyberboss) [inline code/modules/mob/living/carbon/monkey/monkey.dm]: return ..() and this goes at the end

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: config/game_options.txt [hunk 3]
```diff
@@ -495,3 +495,6 @@ MICE_ROUNDSTART 10
 
 ## If the percentage of players alive (doesn't count conversions) drops below this threshold the emergency shuttle will be forcefully called (provided it can be)
 #EMERGENCY_SHUTTLE_AUTOCALL_THRESHOLD 0.2
+
+## Maximum amount of monkeys that can exist. Defaults to 150.
```
Comment (Cyberboss) [inline config/game_options.txt]: update the comment too

**Review by Cyberboss — DISMISSED** [review]

File: code/modules/food_and_drinks/food/snacks_meat.dm [hunk 4]
```diff
@@ -184,6 +184,9 @@
 	foodtype = MEAT | SUGAR
 
 /obj/item/reagent_containers/food/snacks/monkeycube/proc/Expand()
+	if(CONFIG_GET(number/monkeycap) <= GLOB.monkeys.len)
+		visible_message("<span class='warning'>[src] gives a single shudder, before drying out.</span>")
```
Comment (KorPhaeron) [inline code/modules/food_and_drinks/food/snacks_meat.dm]: There should be some indication as to why, even if it wont make perfect sense in universe

**Review by imtakingabreakdontatme — CHANGES_REQUESTED** [review]

**Comment by AnonymousNow** [comment]
Question stands, important for xenobio - does it include dead monkeys?

File: code/_globalvars/lists/mobs.dm [hunk 5]
```diff
@@ -24,3 +24,6 @@ GLOBAL_LIST_EMPTY(spidermobs)				//all sentient spider mobs
 
 GLOBAL_LIST_EMPTY(language_datum_instances)
 GLOBAL_LIST_EMPTY(all_languages)
+
+//Global list of monkeys, used for monkeycap.
+GLOBAL_LIST_EMPTY(monkeys)
```
Comment (optimumtact) [inline code/_globalvars/lists/mobs.dm]: think about all the things you use this list for, then take a good long hard look in the mirror and think about your life choices up to this point.

then when you've done all that, turn this into a counter that is incremented and decremented on monkey intialize and Destroy/Death

**Review by optimumtact — CHANGES_REQUESTED** [review]

File: code/_globalvars/lists/mobs.dm (hunk 5 above)
Comment (MoreRobustThanYou) [reply code/_globalvars/lists/mobs.dm]: fixed

**Review by ChangelingRain — CHANGES_REQUESTED** [review]
If there's something that can produce too many monkeys, nerf that instead.

**Comment by MoreRobustThanYou** [comment]
@ChangelingRain the easiest ways to mass produce monkey cubes are botany and xenobio...

However, it'd be hard to nerf the production of monkey cubes without nerfing the good things, too. It's easy to mass produce shittons of biomass, which can be used for monkey cubes or really fun botany things...

Also monkey cubes are kinda needed for xenobio

File: code/modules/food_and_drinks/food/snacks_meat.dm [hunk 6]
```diff
@@ -184,10 +184,13 @@
 	foodtype = MEAT | SUGAR
 
 /obj/item/reagent_containers/food/snacks/monkeycube/proc/Expand()
+	if(CONFIG_GET(number/monkeycap) <= GLOB.monkeys)
+		visible_message("<span class='warning'>[src] gives a single shudder, before drying out.</span>")
```
Comment (KorPhaeron) [inline code/modules/food_and_drinks/food/snacks_meat.dm]: My comment vanished but you still need to indicate to players why this is happening

**Review by imtakingabreakdontatme — CHANGES_REQUESTED** [review]

**Comment by ChangelingRain** [comment]
If botany can mass-produce them, then they cost too little. Xenobio is also a problem, but a different one, since they use literally all of their monkey production.

File: code/modules/mob/living/carbon/monkey/monkey.dm [hunk 7]
```diff
@@ -35,6 +47,18 @@
 	create_dna(src)
 	dna.initialize_dna(random_blood_type())
 
+/mob/living/carbon/monkey/Destroy()
```
Comment (kevinz000) [inline code/modules/mob/living/carbon/monkey/monkey.dm]: So you can make monkeys negative by killing and then gibbing?

File: code/modules/mob/living/carbon/monkey/monkey.dm [hunk 8]
```diff
@@ -15,9 +15,21 @@
...
+	var/is_in_cap = FALSE //so we don't subtract 2 from GLOB.monkeys
 
 
-/mob/living/carbon/monkey/Initialize()
+
+/mob/living/carbon/monkey/Initialize(mapload, var/ignore_monkeycap = FALSE)
+	if(!ignore_monkeycap && !ckey && CONFIG_GET(number/monkeycap) <= GLOB.monkeys)
+		var/turf/T = get_turf(src)
+		message_admins("Monkey failed to spawn due to monkeycap at [COORD(T)][ADMIN_JMP(T)]!")
+		log_game("Monkey failed to spawn due to monkeycap at [COORD(T)]")
```
Comment (kevinz000) [inline code/modules/mob/living/carbon/monkey/monkey.dm]: Possible spam vector .

**Comment by kevinz000** [comment]
I don't like how this is done tbh.
People can grief by spamming monkies in a room to prevent them from ever spawning.
Shouldn't this be an administrative issue..?

**Comment by AnonymousNow** [comment]
Possible solution: monkey heatmap.

**Comment by Bawhoppen** [comment]
Admin issue

**Comment by KorPhaeron** [comment]
Stale

**Pull request closed without merging**